README.md
[ci-img]: https://img.shields.io/travis/ciena-frost/ember-frost-fields.svg "Travis CI Build Status"
[ci-url]: https://travis-ci.org/ciena-frost/ember-frost-fields

[cov-img]: https://img.shields.io/coveralls/ciena-frost/ember-frost-fields.svg "Coveralls Code Coverage"
[cov-url]: https://coveralls.io/github/ciena-frost/ember-frost-fields

[npm-img]: https://img.shields.io/npm/v/ember-frost-fields.svg "NPM Version"
[npm-url]: https://www.npmjs.com/package/ember-frost-fields

[ember-observer-badge]: http://emberobserver.com/badges/ember-frost-fields.svg "Ember Observer score"
[ember-observer-badge-url]: http://emberobserver.com/addons/ember-frost-fields

[ember-img]: https://img.shields.io/badge/ember-2.3+-orange.svg "Ember 2.3+"

[bithound-img]: https://www.bithound.io/github/ciena-frost/ember-frost-fields/badges/score.svg "bitHound"
[bithound-url]: https://www.bithound.io/github/ciena-frost/ember-frost-fields

# ember-frost-fields

###### Dependencies

![Ember][ember-img]
[![NPM][npm-img]][npm-url]

###### Health

[![Travis][ci-img]][ci-url]
[![Coveralls][cov-img]][cov-url]

###### Security

[![bitHound][bithound-img]][bithound-url]

###### Ember Observer score
[![EmberObserver][ember-observer-badge]][ember-observer-badge-url]

## Overview
Ember-frost-fields provides different kinds of custom input fields and input areas.

### frost-url-input
Based on the [frost-text](http://ciena-frost.github.io/ember-frost-core/#/field) component
from [ember-frost-core](https://github.com/ciena-frost/ember-frost-core), the URL input
field consists of an input field and a button.
The component is used to verify that the provided URL is both valid and accessible.

After clicking the 'Test' button,
+ the entered URL is loosely verified to be correctly formatted.
+ An AJAX GET request is then sent (with JSONP to overcome cross-domain problems) to the destination URL.
+ The response is checked with any status code between 100 and 500, except 404, deemed a success.

Should MIME Checking be enabled and the Type set to 'Text', an error will be thrown when using the Chrome browser

## Testing with ember-hook

### frost-url-input
The file picker component is accessible using ember-hook with the top level hook name or you can access the internal
components as well -
* Default top level hook - `$hook('url-input')`
* Test button hook - `$hook('<hook-name>-button')`
* Input field hook - `$hook('<hook-name>-input')`

## Examples

API and example usage can be found in the sample application in tests/dummy, which is also running at http://ciena-frost.github.io/ember-frost-fields

### frost-url-input

```handlebars
{{frost-url-input}}
```

## Installation

* `git clone` this repository
* `npm install`
* `bower install`

## Running

* `ember server`
* Visit your app at http://localhost:4200.

## Running Tests

* `npm test` (Runs `ember try:testall` to test your addon against multiple Ember versions)
* `ember test`
* `ember test --server`


CHANGELOG.md
# 9.0.0 (2018-04-13)
* **Upgraded** `ember-frost-core` to `^8.0.0`
* **Installed** `ember-cli-svgstore`
* **Updated** frost-icon functionality to work with new version of `ember-frost-core`

# 8.0.0 (2018-03-21)
* **Updated** pull request template
* **Added** issue template
* **Updated** to `pr-bumper` version `3`
* **Updated** to node 8
* **Added** slack integration
* **Updated** `ember-frost-test` to `^4.0.1`
* **Updated** `ember-test-utils` to `^8.1.1`
* **Updated** `ember-cli-frost-blueprints` to `^5.0.2`
* **Updated** `ember-prop-types` to `^7.0.1`
* **Updated** `ember-frost-core` to `^7.0.0`
* **Removed** ignoring of `package-lock.json` file
* **Added** `package-lock.json` file
* **Updated** Travis CI scripts to allow non-exact node version

# 7.0.0 (2018-01-11)
* **Updated** `ember-frost-test` to `^4.0.0`
* **Added** `ember-sinon` @ `^0.7.0`
* **Updated** `ember-test-utils` to `^8.1.0`
* **Updated** `ember-cli-frost-blueprints` to `^5.0.1`
* **Added** ignoring of `package-lock` until we are ready to move to node 8
* **Removed** useLintTree ember-cli-mocha configuration from `ember-cli-build.js`
* **Added** ignoring of linting `CHANGELOG.md`
* **Removed** `.remarkrc` file since it is now provided by `ember-test-utils`
* **Added** `ember-browserify` @ `^1.2.0`
* **Updated** pin `ember-code-snippet` to `1.7.0`
* **Updated** `ember-cli-sass` to `7.1.1`
* **Removed** unused `ember-computed-decorators` package
* **Updated** `ember-frost-core` to `^5.1.1`
* **Updated** `ember-hook` to `1.4.2` and moved to a dependency instead of a devDependency
* **Updated** `ember-prop-types` to `^6.0.0` and moved to a dependency instead of a devDependency
* **Removed** unused `ember-spread` package
* **Removed** unused `ember-truth-helpers` package
* **Removed** unused `ember-concurrency` package
* **Removed** unused `ember-elsewhere` package
* **Removed** blueprint files since packages are now provided via dependencies

# 6.0.2 (2017-12-13)
* Change semver range of `ember-resolver` to align with other repos

# 6.0.1 (2017-11-08)
* Use the latest minor of `ember-frost-core`


# 6.0.0 (2017-11-03)

## WARNING: THIS REVERTS EMBER CLI 2.16.2 BACK TO 2.12.3

We apologize for this change. Unfortunately, due to the internal needs of our organization this became a required action.

The 2.16.2 changes are now located in the `ember-cli-2.16.2` branch and will hopefully be contained in a versioned release again in the future.

# 5.0.0 (2017-10-31)
* **Updated** to Ember CLI 2.16.2 and babel 6
* **Removed** ember-computed-decorators since it is now provided as a dependency of `ember-frost-core`
* **Updated** dependencies
* **Updated** pr-bumber to version 3
* **Updated** CONTRIBUTING.md file
* **Updated** to using Ember Javascript Modules API https://github.com/ember-cli/ember-rfc176-data
* **Updated** blueprints to latest versions of dependencies
* **Updated** to use chrome headless in Travis CI
* **Updated** to using Node 8.1.2 NPM 5
* **Added** eslint-plugin-ember to enforce Ember Javascript Modules API syntax
* **Removed** running of code coverage until issue is resolved with ember-cli-code-coverage: https://github.com/kategengler/ember-cli-code-coverage/issues/133
* **Removed** running of ember-try its-2.12 scenario until issue is resolved: https://github.com/ember-cli/ember-try/issues/148

# 4.2.10 (2017-10-26)
* **Updated** testing dependencies
* **Removed** ember-cli-visual-acceptance
* **Removed** no longer used `bower.json` and `.bowerrc`


# 4.2.9 (2017-08-11)
* Upgrade ember-cli 2.12.3 inter-dependencies

# 4.2.8 (2017-07-05)
* Upgrade `ember-cli` to `2.12.3`

# 4.2.7 (2017-05-10)
* **Updated** the secure tokens in `.travis.yml`

# 4.2.6 (2017-04-21)
* **Added**  blueprint check


# 4.2.5 (2017-03-20)
* **Used** `ember-frost-core` info, warning and error icons instead of the local icons
* **Modified** `ember-cli` dependency to `~` instead of `^` to avoid picking up the latest minor version of ember


# 4.2.4
* **Added** .gitkeep file to the coverage directory

# 4.2.3
* **Updated** the travis scripts used for bumping and publishing

# 4.2.2

* **Fixed** build to publish.

# 4.2.1
* **Updated** integration tests to use the `ember-test-utils` helpers


# 4.2.0

* **Removed** lodash from dependencies.
* **Updated** CI to test in Chrome as well as Firefox.


# 4.1.1

* **Upgraded** to test against Ember 2.11.


# 4.1.0

* **Added** additional builds to CI to make sure addon works with latest versions of Ember.
* **Removed** files from npm package that aren't necessary (all of the various config files).
* **Updated** dependencies to latest versions.


# 4.0.0
- Updated comber-frost-core
- SVG's folder was added to main directory
- 'frost-nav' icon pack was added to package.json
- updated failing dependencies
- removed unnecessary dependencies.

# 3.0.1
**fix** gh-pages demo



# 3.0.0
**updated** publish script perms, coverage attrs, node version.


# 2.0.0
**upgraded** to Ember 2.8.
**Added** sass and md linting.
**Updated** Demo, README and tests

# 1.0.0

* Bumping to 1.0 for inclusion in the long term Frost ecosystem

# 0.2.1

* **Removed** application template from `app` directory which was being used by apps not defining their own
application template.

# 0.2.0

* **Upgraded** from `lodash` 3.x to 4.x.

# 0.1.4
No CHANGELOG section found in Pull Request description.
Use a `# CHANGELOG` section in your Pull Request description to auto-populate the `CHANGELOG.md`

# 0.1.3
No CHANGELOG section found in Pull Request description.
Use a `# CHANGELOG` section in your Pull Request description to auto-populate the `CHANGELOG.md`

# 0.1.2
No CHANGELOG section found in Pull Request description.
Use a `# CHANGELOG` section in your Pull Request description to auto-populate the `CHANGELOG.md`

# 0.1.1
No CHANGELOG section found in Pull Request description.

addon/components/frost-url-input.js
/*
 Based on the textarea field, the URL input field consists of an input field and a button.
 The component is used to verify that the provided URL is both valid and accessible.

 After clicking the 'Test' button, the entered URL is loosely verified to be correctly formatted.
 An AJAX GET request is then sent (with JSONP to overcome cross-domain problems) to the destination URL.
 The response is checked with any status code between 100 and 500, except 404, deemed a success.

 Should MIME Checking be enabled and the Type set to 'Text',
 an error will be thrown when using the Chrome browser
 */

import Ember from 'ember'
const {computed, run, typeOf} = Ember
import {Component} from 'ember-frost-core'
import {PropTypes} from 'ember-prop-types'

import layout from '../templates/components/frost-url-input'

// Override Lint checking problem for undefined $
/* global $ */

export default Component.extend({
  // == Properties ======================================================

  classNames: ['frost-url-input'],
  layout,

  // == State Properties ======================================================

  propTypes: {
    error: PropTypes.bool,
    errorMessage: PropTypes.string,
    isLoading: PropTypes.bool,
    success: PropTypes.bool,
    undetermined: PropTypes.bool,
    placeholder: PropTypes.string,
    urlFormatError: PropTypes.bool
  },

  getDefaultProps () {
    return {
      error: false,
      hook: 'url-field',
      isLoading: false,
      success: false,
      undetermined: false,
      urlFormatError: false,
      placeholder: 'Enter URL'
    }
  },

  errorMessage: computed('urlFormatError', 'error', 'undetermined', 'success'
    , function () {
      if (this.get('urlFormatError')) {
        return 'URL Format Error'
      }

      if (this.get('error') || this.get('undetermined')) {
        return 'error'
      }

      if (this.get('success')) {
        return 'Success'
      }
    }).readOnly(),

  // == Functions =============================================================

  _testError (e) {
    const props = {
      isLoading: false
    }

    let httpStatusCode = e.status
    // JSONP not supported by server

    if (httpStatusCode === 404) {
      props.error = true
    } else if ((httpStatusCode >= 100) && (httpStatusCode < 500)) {
      props.success = true
    } else {
      props.undetermined = true
    }

    run(() => {
      this.setProperties(props)
    })
  },

  _testStatus () {
    run(() => {
      this.setProperties({
        isLoading: false,
        success: true
      })
    })
  },

  // == Actions ===============================================================

  actions: {
    input (e) {
      const onInput = this.attrs['onInput']

      this.actions.clear.call(this, false)

      if (typeOf(onInput) === 'function') {
        onInput({
          id: this.get('id'),
          value: e.target.value
        })
      }
    },

    test (host) {
      this.send('clear', false)

      // Verify URL - potentially change to a solid regex in the future
      if (!host || (!host.toLowerCase().startsWith('http://') && !host.toLowerCase().startsWith('https://'))) {
        this.set('urlFormatError', true)
        return
      }

      this.set('isLoading', true)

      // Not using Ember AJAX as it throws unrecoverable error when JSONP is not supported by the server
      $.ajax({
        url: host,
        crossDomain: true,
        dataType: 'jsonp text',
        jsonp: false,
        timeout: 10000,
        status: this._testStatus.bind(this),
        error: this._testError.bind(this)
      })
    },

    clear (clearField) {
      if (clearField) {
        this.set('value', '')
      }

      this.setProperties({
        error: false,
        isLoading: false,
        success: false,
        undetermined: false,
        urlFormatError: false
      })
    }
  }
})
